llvm/labs/lab2/kruglov_alexey/MulToBitShift.cpp

```diff
@@ -0,0 +1,82 @@
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include <stack>
+
+namespace {
+struct MulToBitShift : llvm::PassInfoMixin<MulToBitShift> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &FAM) {
+    for (llvm::BasicBlock &BB : F) {
+      std::stack<llvm::Instruction *> worklist;
+      for (llvm::Instruction &Inst : BB) {
+        if (Inst.getOpcode() == 17) { // Opcode for 'mul' Inst is 17
+          llvm::BinaryOperator *op =
+              llvm::dyn_cast<llvm::BinaryOperator>(&Inst);
+          llvm::IRBuilder<> builder(op);
+          llvm::Value *lhs = op->getOperand(0);
+          llvm::Value *rhs = op->getOperand(1);
+          int lg1 = -2, lg2 = -2;
+          if (llvm::ConstantInt *CIl = llvm::dyn_cast<llvm::ConstantInt>(lhs)) {
+            lg1 = CIl->getValue().exactLogBase2();
+          }
+          if (llvm::ConstantInt *CIr = llvm::dyn_cast<llvm::ConstantInt>(rhs)) {
+            lg2 = CIr->getValue().exactLogBase2();
+          }
+          if (lg1 < lg2) {
+            std::swap(lg1, lg2);
+            std::swap(lhs, rhs);
+          }
+          if (lg1 > -1) {

```

Comment:
I think it is hardly can be <0, but to be more accurate you can add an assert(lg1 >= 0) here.

Reply:
It can be -2, if the operand is not constantInt, -1 if operand is not 2 in integer power, and >=0 if it is, then lg is the power of 2